### ISSUE TITLE
Completion of record labels useless under pattern matching

### ISSUE THREAD

**Issue by choeger** [the-lambda-church/merlin#585]
When I select an element of a module, I get a satisfying completion offer.  Records, however, are way too broad:

![merlin](https://cloud.githubusercontent.com/assets/179956/19362547/6761e258-9187-11e6-84f2-c96ad2e5e311.png)

It does contain the correct field, but also so many others that the completion is barely useful.

As a related request: Why are there **Module** completions offered?

**Comment by let-def** [comment]
Yes, completion of records is syntactic, Merlin doesn't know the type of the record you are trying to complete. Though there is a prototype with that.

It shows:
1. all fields/labels in scope, because that's what you can project from a record. But it cannot filter labels with incorrect type.
2. all modules, because if your record is defined in a different module you need to quality the path (or rely on disambiguation, which is not syntactic either). For instance, projecting `pos_cnum` when `Lexing` module is not in scope is written `my_pos.Lexing.pos_cnum`.

**Comment by ygrek** [comment]
It looks strange that in cases when merlin knows type of record precisely (ie getting type of value works and is correct and drilling down into the type lists all fields correctly) it doesn't propose correct fields for completion.

There is now bounty on this issue https://www.bountysource.com/issues/39169510-completion-of-record-labels-useless-under-pattern-matching

**Comment by ygrek** [comment]
I see that example above works alright (except it still shows useless modules, but proposed label is only one and correct). But in bigger project it exhibits reported problem. I will try to add repro.

**Comment by ygrek** [comment]
#FAIL 
it actually works alright when pinned to dev-repo
Would be nice to not show the list of irrelevant modules for the ideal interaction though :)

**Comment by ygrek** [comment]
@let-def for the purposes of bounty I consider this feature done, so please feel free to claim it before it expires and bountysource gets the money (ref https://help.nextcloud.com/t/dont-use-bountysource-anymore/84943 )
